products/bluebell/src/lexer.rs

```diff
@@ -0,0 +1,573 @@
+                Token::Equals => "=",
+                Token::Ampersand => "&",
+                Token::LeftArrow => "<-",
+                Token::ColonEquals => ":=",
+                Token::At => "@",
+                Token::Minus => "-",
+                Token::Underscore => "_",
+                Token::Forall => KEYWORD_FORALL,
+                Token::Builtin => KEYWORD_BUILTIN,
+                Token::Library => KEYWORD_LIBRARY,
+                Token::Import => KEYWORD_IMPORT,
+                Token::Let => KEYWORD_LET,
+                Token::In => KEYWORD_IN,
+                Token::Match => KEYWORD_MATCH,
+                Token::With => KEYWORD_WITH,
+                Token::End => KEYWORD_END,
+                Token::Fun => KEYWORD_FUN,
+                Token::Tfun => KEYWORD_TFUN,
+                Token::Contract => KEYWORD_CONTRACT,
+                Token::Transition => KEYWORD_TRANSITION,
+                Token::Send => KEYWORD_SEND,
+                Token::Field => KEYWORD_FIELD,
+                Token::Accept => KEYWORD_ACCEPT,
+                Token::Exists => KEYWORD_EXISTS,
+                Token::Delete => KEYWORD_DELETE,
+                Token::Throw => KEYWORD_THROW,
+                Token::Map => KEYWORD_MAP,
+                Token::ScillaVersion => KEYWORD_SCILLA_VERSION,
+                Token::Type => KEYWORD_TYPE,
+                Token::Of => KEYWORD_OF,
+                Token::As => KEYWORD_AS,
+                Token::Procedure => KEYWORD_PROCEDURE,
+                Token::Emp => KEYWORD_EMP,
+                Token::Event => KEYWORD_EVENT,
+                Token::EventType => KEYWORD_EVENT_TYPE,
+                Token::ByStr => KEYWORD_BYSTR,
+
+
+                Token::Whitespace => " ",
+                _ => "?", // Token::Unknown made as a wild card to avoid compiler complaining.
+            }
+            .to_string(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ParseError {
+    // Not possible
+}
+
+/// Provides the ability to tokenize a source file.
+pub struct Lexer<'input> {
+    /// An iterator of the source file's characters, along with their indices.
+    chars: Peekable<CharIndices<'input>>,
+    /// A reference to the source file being tokenized.
+    document: &'input str,
+    /// The current line number being tokenized.
+    line: usize,
+    /// The current character number within the current line being tokenized.
+    character: usize,
+}
+
+impl<'input> Lexer<'input> {
+    pub fn new(input: &'input str) -> Self {
+        Lexer {
+            chars: input.char_indices().peekable(),
+            document: input,
+            line: 1,
+            character: 1,
+        }
+    }
+}
+
+impl<'input> Iterator for Lexer<'input> {
+    type Item = Spanned<Token<&'input str>, usize, ParseError>;
+
+    // <(usize, Token, usize, usize, usize);
+
+    fn next(&mut self) -> Option<Self::Item> {
+
+
+        while let Some((start, ch)) = self.chars.next() {
```

Comment:
Are all the tokens described by a regular language? I wonder if we could simplify the parser by describing the production of each token by a regular expression instead. I'll do a bit of research to see if there are any convenient solutions..

Reply:
I did use solely regexes inside lalrpop to describe the lexer to start with. The problem that occurred is that lalrpopd does not have support for non-greedy expressions and for that reason it was not possible to create pattern that would match comments.